controller-server/src/main/java/com/yahoo/vespa/hosted/controller/persistence/ApplicationSerializer.java

```diff
@@ -258,15 +258,13 @@ private void zoneIdToSlime(ZoneId zone, Cursor object) {
     }
 
     private void toSlime(ApplicationVersion applicationVersion, Cursor object) {
-        if (applicationVersion.buildNumber().isPresent() && applicationVersion.source().isPresent()) {
```

Comment:
@jonmv This was the bug. When source revision is empty, no fields are written to the `latestVersion` object, but the object itself is created, thus becoming an unknown version. Subsequent reads fail because the `Application` constructor throws if `latestVersion` is unknown so fromSlime/toSlime is asymmetric.

> What are the other changes here?

Cleanup of `SourceRevision`. `SourceRevision` where all fields are empty is now treated as an empty source revision.

Reply:
Yes, this was the bug. We still have serialised `unknown` `latestVersion`s, though? Or did you clean them out by hand?

Are the fields all empty? I'd imagine the `commit` field wasn't. The plan was to remove `SourceRevision`, since they're only used for the `commit` and the generated source URLs.

Reply:
>  Or did you clean them out by hand?

Yes.

> Are the fields all empty?

In the production data, `sourceRevision` is either a) `null`, b) an object with all non-empty or fields or c) an object with all empty fields.

Reply:
> The plan was to remove SourceRevision.

Please do! 🔥